### DISCUSSION TITLE
Upgrading Seqr

### DISCUSSION THREAD

**Discussion by atb1r21** [broadinstitute/seqr#2609] [Q&A]
Hi, I was running the previous versions of seqr via docker-compose on a VM at my institution(last updated circa 2021), however as a result of the post at https://github.com/broadinstitute/seqr/discussions/2501 I have attempted to upgrade this as advised.

This didn't work and I am seeing issues where the kibana container is always unhealthy. Additionally I could eventually get the seqr site page to open for my local installation but users could no longer log in.

I have been left limited documentation and I think I need some help regarding upgrading this, as it doesn't seem to be as smooth for my setup as the discussion page linked above would suggest. This is particularly true as when I first tried to upgrade the postgres container wouldn't start again which is obviously a large issue.

If any support or help could be rendered here I would be very grateful, I have resorted to having my VM restored from a backup, but additionally once restored the older version of seqr will not start either due to a python error.

I am happy to discuss further and provide any outputs/screenshots necessary but am coming here to look for a first point of contact in resolving this issue, as the researchers using seqr are keen to have it back in action.

Looking forward to hearing from you,
Alister

**Comment by hanars** [comment]
I'm so sorry you are having this experience. Can you please provide the command(s) you ran to perform the update, any output you have, and also a copy of the docker-compose file you are currently using

**Reply by atb1r21** [reply]
Hi,

I have attached the current docker-compose file as a .txt. All I did to perform the update was stop the current docker containers, and then the following instructions via #2501 

Download a new version of the [docker-compose.yml](https://raw.githubusercontent.com/broadinstitute/seqr/master/docker-compose.yml) file as described in the [local install documentation]

(https://github.com/broadinstitute/seqr/blob/master/deploy/LOCAL_INSTALL.md)

Pull updated docker images with the “docker-compose pull” command

Restart seqr with the “docker-compose up -d seqr” command

[docker-compose.txt](https://github.com/broadinstitute/seqr/files/8408631/docker-compose.txt)

The current state of my seqr docker images are:

CONTAINER ID   IMAGE                                                         COMMAND                  CREATED        STATUS                  PORTS                                                 NAMES
90be576ed683   gcr.io/seqr-project/pipeline-runner:gcloud-prod               "/entrypoint.sh"         10 days ago    Up 10 days                                                                    srv_pipeline-runner_1
f2fe5203104c   gcr.io/seqr-project/kibana:gcloud-prod                        "/usr/local/bin/dumb…"   10 days ago    Up 2 days (unhealthy)   5601/tcp                                              srv_kibana_1
164ff5cbfdb7   gcr.io/seqr-project/seqr:gcloud-prod                          "/entrypoint.sh"         10 days ago    Up 2 days (healthy)     0.0.0.0:80->8000/tcp, :::80->8000/tcp                 srv_seqr_1
2e39a197686b   gcr.io/seqr-project/elasticsearch:gcloud-prod-elasticsearch   "/tini -- /usr/local…"   7 months ago   Up 10 days (healthy)    0.0.0.0:9200->9200/tcp, :::9200->9200/tcp, 9300/tcp   elasticsearch
6e54b52333b8   gcr.io/seqr-project/redis:gcloud-prod                         "docker-entrypoint.s…"   7 months ago   Up 10 days (healthy)    6379/tcp                                              srv_redis_1
ce2585f1b7fa   gcr.io/seqr-project/postgres:gcloud-prod                      "docker-entrypoint.s…"   7 months ago   Up 10 days (healthy)    5432/tcp  

My main question is that when I first attempted the upgrade that srv_postgres_1 would not start as the data folder was not empty, but I could start it by creating an empty data directory and pointing it at that. I got it to the point where users could access seqr via a browser, but they could no longer log in and when I created a super user account, there was no list of users.

From what I can see in the newest docker-compose.yaml file that I am meant to use to upgrade, some of the container images have moved but not much else has changed, which is why I am a bit lost as to why I am seeing so many issues.

Thanks,
Alister

**Reply by sjahl** [reply]
Hi Alister, I'm sorry this happened to you! The changes in #2501 should only have affected the seqr docker image -- I'm not sure what the circumstances might be that would have introduced changes to your postgres container. That postgres failure seems like the biggest presenting problem to me, so I'd like to focus on what's going on there. Do you have any log messages from that first attempted upgrade where postgres failed to restart? It would be helpful to me to see all the output that you received from docker-compose at the time.

In the interest of getting your old/restored seqr back up and running, what is the python error you are receiving when trying to start seqr? Are there any other log messages being output prior to the python error? Was your postgres container also restored from your backup to the same point in time?

**Reply by sjahl** [reply]
Ah, I see that our new docker-compose file switched to the upstream `postgres` image at some point, and the docker-compose.txt file you've posted uses `gcr.io/seqr-project/postgres:gcloud-prod` that we used to publish. While waiting for additional log output, I'll see if I can track down what upgrading from one to the other looks like in a test environment.

**Reply by atb1r21** [reply]
Hi,

Unforunately no, I was concerned that I had corrupted the database or similar and so I have rolled it back, I didn't make notes at the time, as I was just testing out various things regarding docker.

As a note I would say that it looks like not only the seqr image is affected.

Our previous docker file has the following postgres configuration:

  postgres:
    **image: gcr.io/seqr-project/postgres:gcloud-prod**
    environment:
      - PGPORT=5433
      - POSTGRES_PASSWORD=docker-compose-postgres-password
    volumes:
      - ./data/postgres:/var/lib/postgresql/data
    healthcheck:
      test: pg_isready -h postgres -U postgres
      interval: 5s
      timeout: 10s
      retries: 100

Whilst the new one has the folllowing:

  postgres:
    **image: postgres:12.6**
    command: postgres -c listen_addresses='*'
    environment:
      - PGPORT=5433
      - POSTGRES_PASSWORD=docker-compose-postgres-password
    volumes:
      - ./data/postgres:/var/lib/postgresql/data
    healthcheck:
      test: pg_isready -h postgres -U postgres
      interval: 5s
      timeout: 10s
      retries: 100

Is there a change in image here? I'm seeing a few changes in the image path in the new docker-compose.yaml compared to our old one.

**Reply by atb1r21** [reply]
We seem to have posted at the same time. Yes, I think that's part of the issue. To give you logs I will need to attempt the upgrade again, is there any chance the actual data in the database can be overwritten or destroyed?

**Reply by sjahl** [reply]
I don't think there's anything in the postgres startup that will delete data, but there may be a postgres version upgrade, depending on how old that postgres:gcloud-prod image from our gcr registry was, so I would ensure that you've taken a good backup before proceeding. If you'd like to hold off, I'm happy to try and reproduce what you're seeing here, to see if I can identify a working upgrade path.

**Reply by hanars** [reply]
> My main question is that when I first attempted the upgrade that srv_postgres_1 would not start as the data folder was not empty, but I could start it by creating an empty data directory and pointing it at that

There is nothing in startup that deletes data, but it is important that the data directory used by postgres is maintained. Pointing to an empty data directory would have the effect you describe of having none of your users or records in seqr exist. Assuming you did not actually delete the old data directory, can you make sure your local `./data/postgres` is not empty, copy that directory somewhere (as a backup just in case something goes wrong ) and then change your configuration back from the empty folder to that folder and provide us the logs when you try to start it up again

**Reply by sjahl** [reply]
@atb1r21 If your error looks like the following:

```
seqrtest-postgres-1  | initdb: error: directory "/var/lib/postgresql/data" exists but is not empty
seqrtest-postgres-1  | If you want to create a new database system, either remove or empty
seqrtest-postgres-1  | the directory "/var/lib/postgresql/data" or run initdb
seqrtest-postgres-1  | with an argument other than "/var/lib/postgresql/data".
```

Then, it appears that the default location for postgres to house its data has changed between the two container versions. I believe the solution is to tell postgres what the path to the old folder was, which can be done by specifying the `PGDATA` variable in the postgres container's environment. e.g.:

```
  postgres:
    image: postgres:12.6
    command: postgres -c listen_addresses='*'
    environment:
      - PGPORT=5433
      - POSTGRES_PASSWORD=docker-compose-postgres-password
      - PGDATA=/var/lib/postgresql/data/pgdata
```

My understanding of the difference is that the default value in the newer container is to set PGDATA to /var/lib/postgresql/data, whereas it used to be /var/lib/postgresql/data/pgdata. This causes the postgres database container to refuse to start, because the database structure no longer looks the same as what it is expecting when looking at the correct folder. I was able to get the database to restart cleanly in my test env by adding that PGDATA variable set to the old directory location.

@hanars It's kind of hard to tell what to do here on our end. Anyone who's started seqr after we deprecated the old postgres images won't have to set this, and anyone who's got an older installation will. So, changing the value in our official compose file will break people either way.

**Reply by hanars** [reply]
I think we should leave it as is so that people with newer installations will work and people with older installations will need to update. I'm not sure what to do for the people with older installations other than pinning your answer

**Reply by atb1r21** [reply]
Hi both,

@sjahl That was my error as you've posted, so I'll make a backup and try again.

I'll give that a shot and get back to you. It might be a while as I am on annual leave for the next couple of weeks but should have news soon after that.

I appreciate the hard work here.

Thanks,
Alister

**Comment by atb1r21** [comment]
Hi,

That worked:

9c798d8e90fe   gcr.io/seqr-project/pipeline-runner:gcloud-prod        "/entrypoint.sh"         11 minutes ago   Up 11 minutes                                                                     srv_pipeline-runner_1

81d90b627b18   docker.elastic.co/kibana/kibana:7.16.3                 "/bin/tini -- /usr/l…"   12 minutes ago   Up 12 minutes (unhealthy)   5601/tcp                                              srv_kibana_1

6c0371d6a7bb   gcr.io/seqr-project/seqr:gcloud-prod                   "/entrypoint.sh"         13 minutes ago   Up 13 minutes (healthy)     0.0.0.0:80->8000/tcp, :::80->8000/tcp                 srv_seqr_1

8570e3d66003   postgres:12.6                                          "docker-entrypoint.s…"   13 minutes ago   Up 13 minutes (healthy)     5432/tcp                                              srv_postgres_1

16de98c59c0f   gcr.io/seqr-project/redis:gcloud-prod                  "docker-entrypoint.s…"   13 minutes ago   Up 13 minutes (healthy)     6379/tcp                                              srv_redis_1

959bc9f61fa8   docker.elastic.co/elasticsearch/elasticsearch:7.16.3   "/bin/tini -- /usr/l…"   13 minutes ago   Up 13 minutes (healthy)     0.0.0.0:9200->9200/tcp, :::9200->9200/tcp, 9300/tcp   elasticsearch

The only issue that remains is that the kibana container is unhealthy. Regardless of the version and other settings I have never seen the kibana container running in a healthy state, is there an issue here that I need to resolve on my end, the logs for the kibana container are not very indicative.

I have asked our users to test login to the web front end of seqr and get back to me.
